### ISSUE TITLE
[QUESTION]: data loss in our production env

### ISSUE THREAD

**Issue by jjeffcaii** [dgraph-io/badger#1991]
### Question.

We use badger as our embed storage in an online service, the codes which initialize the badger DB is below:
```golang
    // github.com/dgraph-io/badger/v3 v3.2103.5
    opt := badger.DefaultOptions(path).
		WithLogger(badgerLogger{}).
		WithNumVersionsToKeep(1).
		WithNumCompactors(2).
		WithMemTableSize(32 << 20).
		WithNumMemtables(1).
		WithNumLevelZeroTables(1).
		WithNumLevelZeroTablesStall(2).
		WithCompactL0OnClose(true).
		WithValueThreshold(1 << 10).
		WithCompression(options.None).
		WithBlockCacheSize(0).
		WithIndexCacheSize(0)
    badger.Open(opt)
```

And also we start a timer to run GC:
```golang
	// https://dgraph.io/docs/badger/get-started/#garbage-collection
	defer b.gcTicker.Stop()
	for range b.gcTicker.C {
	again:
		err := b.db.RunValueLogGC(0.7)
		if err == nil {
			goto again
		}
		logger.Info("~~~ run badger gc ok ~~~")
	}
```

And some day we met a problem, we cannot query some rows, it will report an error like below:
```text
2023/08/07 13:01:43 [ERROR] Key: [103 114 97 58 0 0 0 0 0 0 0 5], Version : 4, meta: 2, userMeta: 0 valuePointer: {Fid:6 Len:1361 Offset:979793}
```
It seems the target key is found in SST file `543.sst` at Level-5, with version 4, and it points to the `000006.vlog`, but I there's no `000006.vlog` already.

And I checked all SST files, there's another `564.sst` at Level-6, with same version, but points to the `000007.vlog`, this should be the right target files, the `000007.vlog` is really exist, and I wrote some codes to decode it manually, the lost rows are in this file.

In short words:
<img width="1040" alt="image" src="https://github.com/dgraph-io/badger/assets/1284892/e6fc8098-b8cd-4262-8e55-3848c389e06c">

I am not familiar with the internal logic of badger, this is a small probability problem. Can anyone help? Thanks!

**Comment by kingljl** [comment]
I encountered the same problem

**Comment by jjeffcaii** [comment]
@mangalaman93 How is the issue going? I can provide some db snapshot files if it can help to diagnose the problem.

**Comment by jjeffcaii** [comment]
@mangalaman93 I've emailed you the data snapshot files, the vlog file was removed because of some security reason, and hope it's helpful for diagnosing the problem, thanks.

**Comment by jjeffcaii** [comment]
It seems that our application occurs core-dump when compaction is executing, that broke the SST files, I think.
I'll close the issue. Thanks all.
